src/Microsoft.Health.Dicom.Core/Features/Indexing/ITagReindexOperationStore.cs

```diff
@@ -0,0 +1,57 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Health.Dicom.Core.Features.ExtendedQueryTag;
+
+namespace Microsoft.Health.Dicom.Core.Features.Indexing
+{
+    /// <summary>
+    /// Store that records relationship between extended query tag and reindex operation.
+    /// </summary>
+    public interface ITagReindexOperationStore
+    {
+        /// <summary>
+        /// Get entires of operations.
+        /// </summary>
+        /// <param name="operationId">The operation id.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The entries.</returns>
+        Task<IReadOnlyList<TagReindexOperationEntry>> GetEntriesOfOperationAsync(
+            string operationId,
+            CancellationToken cancellationToken = default);
+
+        /// <summary>
+        ///  Get next watermarks for operation.
+        /// </summary>
+        /// <param name="operationId">The operation id.</param>
+        /// <param name="count">The count of watermarks</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>Watermark list.</returns>
+        Task<IReadOnlyList<long>> GetNextWatermarksOfOperationAsync(string operationId, int count, CancellationToken cancellationToken = default);
```

Comment:
What is this method doing? It seems like it should't have to worry about "count" or returning a list

Reply:
The count will be `BatchSize` * `MaxParallelBatches`.
this method call a storage procedure which return next set of watermarks we need to proceed. The number of watermarks is defined by `count`

Reply:
This is because we're returning the exact watermarks right now instead of the batches?

Reply:
I think I'm using a different way from your design. we could talk about this tomorrow.